// oceanic.js/dist/lib/structures/Application.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const tslib_1 = require("tslib");
/** @module Application */
const ClientApplication_1 = tslib_1.__importDefault(require("./ClientApplication"));
const OAuthGuild_1 = tslib_1.__importDefault(require("./OAuthGuild"));
const User_1 = tslib_1.__importDefault(require("./User"));
const Team_1 = tslib_1.__importDefault(require("./Team"));
const Routes = tslib_1.__importStar(require("../util/Routes"));
/** Represents an application. */
class Application extends ClientApplication_1.default {
    /** The approximate number of guilds the application is in. */
    approximateGuildCount;
    /** The approximate number of users this application has been installed by. */
    approximateUserInstallCount;
    /** If the bot can be invited by anyone. */
    botPublic;
    /** If authorizing the bot requires a code grant. */
    botRequireCodeGrant;
    /** This application's rich presence invite cover image hash, if any. */
    coverImage;
    /** This application's default custom authorization link, if any. */
    customInstallURL;
    /** The description of the application. */
    description;
    /** The state of this application's discoverability. */
    discoverabilityState;
    /** The { @link Constants~ApplicationDiscoveryEligibilityFlags | flags } for this application's discovery eligibility. */
    discoveryEligibilityFlags;
    /** If webhook events are enabled for the app. */
    eventWebhooksStatus;
    /**	List of Webhook event types the app subscribes to. */
    eventWebhooksTypes;
    /** Event webhooks URL for the app to receive webhook events. */
    eventWebhooksURL;
    /** The explicit content filter for this application. */
    explicitContentFilter;
    /** If this application is a game sold on Discord, the guild to which it has been linked. This will only be present if recieved via {@link REST/Applications.getCurrent | `/applications/@me`}. */
    guild;
    /** If this application is a game sold on Discord, the ID of the guild to which it has been linked. */
    guildID;
    hook;
    /** The icon hash of the application. */
    icon;
    /** Settings for this application's in-app authorization link, if enabled. */
    installParams;
    /** The install types available for this application. */
    integrationTypes;
    /** The configs for the install types available for this application. */
    integrationTypesConfig;
    /** This applications interaction endpoint url, if any. */
    interactionsEndpointURL;
    /** The event types that will be recieved like http interactions, if interactionsVersion is 2. */
    interactionsEventTypes;
    /** The interactions version of this application. */
    interactionsVersion;
    /** If this application is monetized. */
    isMonetized;
    /** The { @link Constants~ApplicationMonetizationEligibilityFlags | flags } for this application's monetization eligibility. */
    monetizationEligibilityFlags;
    /** This application's monetization state. */
    monetizationState;
    /** The name of the application. */
    name;
    /** The owner of this application. */
    owner;
    /** If this application is a game sold on Discord, the id of the Game's SKU. */
    primarySKUID;
    /** A URL to this application's privacy policy. */
    privacyPolicyURL;
    /** The redirect URIs for this application. */
    redirectURIs;
    /** This application's role connections verification url, if any. */
    roleConnectionsVerificationURL;
    /** The state of this application's RPC application. */
    rpcApplicationState;
    /** A list of rpc origin urls, if rpc is enabled. */
    rpcOrigins;
    /** If this application is a game sold on Discord, the slug that links to its store page. */
    slug;
    /** The state of this application's store application state. */
    storeApplicationState;
    /** The tags for this application. */
    tags;
    /** The team that owns this application. */
    team;
    /** A URL to this application's terms of service. */
    termsOfServiceURL;
    /** The type of this application. */
    type;
    /** The state of this application's verification. */
    verificationState;
    /** The bot's hex encoded public key. */
    verifyKey;
    constructor(data, client) {
        super(data, client);
        this.approximateGuildCount = data.approximate_guild_count ?? 0;
        this.approximateUserInstallCount = data.approximate_user_install_count ?? 0;
        this.botPublic = data.bot_public;
        this.botRequireCodeGrant = data.bot_require_code_grant;
        this.coverImage = data.cover_image ?? null;
        this.description = data.description;
        this.discoverabilityState = data.discoverability_state;
        this.discoveryEligibilityFlags = data.discovery_eligibility_flags;
        this.eventWebhooksStatus = data.event_webhooks_status;
        this.eventWebhooksTypes = data.event_webhooks_types;
        this.eventWebhooksURL = data.event_webhooks_url;
        this.explicitContentFilter = data.explicit_content_filter;
        this.guild = data.guild === undefined ? null : new OAuthGuild_1.default(data.guild, client);
        this.guildID = data.guild_id ?? null;
        this.hook = data.hook;
        this.icon = data.icon;
        this.installParams = data.install_params;
        this.integrationTypes = [];
        this.integrationTypesConfig = {};
        this.interactionsEndpointURL = null;
        this.interactionsEventTypes = data.interactions_event_types;
        this.interactionsVersion = data.interactions_version;
        this.isMonetized = data.is_monetized;
        this.monetizationEligibilityFlags = data.monetization_eligibility_flags;
        this.monetizationState = data.monetization_state;
        this.name = data.name;
        this.owner = data.owner === undefined ? null : new User_1.default(data.owner, client);
        this.primarySKUID = data.primary_sku_id;
        this.privacyPolicyURL = data.privacy_policy_url;
        this.redirectURIs = data.redirect_uris ?? [];
        this.roleConnectionsVerificationURL = null;
        this.rpcApplicationState = data.rpc_application_state;
        this.rpcOrigins = data.rpc_origins ?? [];
        this.slug = data.slug;
        this.storeApplicationState = data.store_application_state;
        this.tags = data.tags ?? [];
        this.team = data.team ? new Team_1.default(data.team, client) : null;
        this.termsOfServiceURL = data.terms_of_service_url;
        this.type = data.type;
        this.verificationState = data.verification_state;
        this.verifyKey = data.verify_key;
        this.update(data);
    }
    /**
     * The url of this application's cover image.
     * @param format The format the url should be.
     * @param size The dimensions of the image.
     */
    coverImageURL(format, size) {
        return this.coverImage === null ? null : this.client.util.formatImage(Routes.APPLICATION_COVER(this.id, this.coverImage), format, size);
    }
    /**
     * The url of this application's icon.
     * @param format The format the url should be.
     * @param size The dimensions of the image.
     */
    iconURL(format, size) {
        return this.icon === null ? null : this.client.util.formatImage(Routes.APPLICATION_ICON(this.id, this.icon), format, size);
    }
    toJSON() {
        return {
            ...super.toJSON(),
            approximateGuildCount: this.approximateGuildCount,
            botPublic: this.botPublic,
            botRequireCodeGrant: this.botRequireCodeGrant,
            coverImage: this.coverImage,
            customInstallURL: this.customInstallURL,
            description: this.description,
            discoverabilityState: this.discoverabilityState,
            discoveryEligibilityFlags: this.discoveryEligibilityFlags,
            explicitContentFilter: this.explicitContentFilter,
            guild: this.guild?.toJSON() ?? null,
            guildID: this.guildID,
            hook: this.hook,
            icon: this.icon,
            installParams: this.installParams,
            integrationTypes: this.integrationTypes,
            integrationTypesConfig: this.integrationTypesConfig,
            interactionsEndpointURL: this.interactionsEndpointURL,
            interactionsEventTypes: this.interactionsEventTypes,
            interactionsVersion: this.interactionsVersion,
            isMonetized: this.isMonetized,
            monetizationEligibilityFlags: this.monetizationEligibilityFlags,
            monetizationState: this.monetizationState,
            name: this.name,
            owner: this.owner?.toJSON() ?? null,
            primarySKUID: this.primarySKUID,
            privacyPolicyURL: this.privacyPolicyURL,
            redirectURIs: this.redirectURIs,
            roleConnectionsVerificationURL: this.roleConnectionsVerificationURL,
            rpcApplicationState: this.rpcApplicationState,
            rpcOrigins: this.rpcOrigins,
            slug: this.slug,
            storeApplicationState: this.storeApplicationState,
            tags: this.tags,
            team: this.team?.toJSON() ?? null,
            termsOfServiceURL: this.termsOfServiceURL,
            type: this.type,
            verificationState: this.verificationState,
            verifyKey: this.verifyKey
        };
    }
}
exports.default = Application;
