// oceanic.js/dist/lib/structures/ApplicationCommand.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const tslib_1 = require("tslib");
/** @module ApplicationCommand */
const Base_1 = tslib_1.__importDefault(require("./Base"));
const Permission_1 = tslib_1.__importDefault(require("./Permission"));
const Constants_1 = require("../Constants");
const Errors_1 = require("../util/Errors");
/** Represents an application command. */
class ApplicationCommand extends Base_1.default {
    _cachedGuild;
    /** The application this command is for. */
    application;
    /** The ID of application this command is for. */
    applicationID;
    /** The contexts in which this application command can be used in. */
    contexts;
    /** The default permissions for this command. */
    defaultMemberPermissions;
    /** The description of this command. Empty string for non `CHAT_INPUT` commands. */
    description;
    /** A dictionary of [locales](https://discord.com/developers/docs/reference#locales) to localized descriptions. */
    descriptionLocalizations;
    /** The description of this application command in the requested locale. */
    descriptionLocalized;
    /** If this command can be used in direct messages (global commands only). */
    dmPermission;
    /** The id of the guild this command is in (guild commands only). */
    guildID;
    /** The handler of this command. Only valid for `PRIMARY_ENTRY_POINT` commands. */
    handler;
    /** The installation contexts in which this command is available. */
    integrationTypes;
    /** The name of this command. */
    name;
    /** A dictionary of [locales](https://discord.com/developers/docs/reference#locales) to localized names. */
    nameLocalizations;
    /** The description of this application command in the requested locale. */
    nameLocalized;
    /** Whether the command is age restricted. */
    nsfw;
    /** The options on this command. Only valid for `CHAT_INPUT`. */
    options;
    /** The [type](https://discord.com/developers/docs/interactions/application-commands#application-command-object-application-command-types) of this command. */
    type;
    /** Autoincrementing version identifier updated during substantial record changes. */
    version;
    constructor(data, client) {
        super(data.id, client);
        this.application = client["_application"] && client.application.id === data.application_id ? client.application : undefined;
        this.applicationID = data.application_id;
        this.contexts = data.contexts;
        this.defaultMemberPermissions = data.default_member_permissions ? new Permission_1.default(data.default_member_permissions) : null;
        this.description = data.description;
        this.descriptionLocalizations = data.description_localizations;
        this.descriptionLocalized = data.description_localized;
        this.dmPermission = data.dm_permission;
        this.guildID = data.guild_id ?? null;
        this.handler = data.handler;
        this.integrationTypes = data.integration_types;
        this.name = data.name;
        this.nameLocalizations = data.name_localizations;
        this.nameLocalized = data.name_localized;
        this.nsfw = data.nsfw;
        this.options = data.options?.map(o => client.util.optionToParsed(o));
        this.type = (data.type ?? Constants_1.ApplicationCommandTypes.CHAT_INPUT);
        this.version = data.version;
    }
    /** The guild this command is in (guild commands only). This will throw an error if the guild is not cached. */
    get guild() {
        if (this.guildID !== null && this._cachedGuild !== null) {
            this._cachedGuild ??= this.client.guilds.get(this.guildID);
            if (!this._cachedGuild) {
                throw new Errors_1.UncachedError(this, "guild", "GUILDS", this.client);
            }
            return this._cachedGuild;
        }
        return this._cachedGuild === null ? this._cachedGuild : (this._cachedGuild = null);
    }
    /**
     * Delete this command.
     */
    async delete() {
        return this.guildID ? this.client.rest.applications.deleteGuildCommand(this.applicationID, this.guildID, this.id) : this.client.rest.applications.deleteGlobalCommand(this.applicationID, this.id);
    }
    /**
     * Edit this command.
     * @param options The options for editing the command.
     */
    async edit(options) {
        return this.guildID ? this.client.rest.applications.editGuildCommand(this.applicationID, this.guildID, this.id, options) : this.client.rest.applications.editGlobalCommand(this.applicationID, this.id, options);
    }
    /**
     * Edit this command's permissions (guild commands only). This requires a bearer token with the `applications.commands.permissions.update` scope.
     * @param options The options for editing the permissions.
     */
    async editGuildCommandPermissions(options) {
        if (!this.guildID) {
            throw new TypeError("editGuildCommandPermissions cannot be used on global commands.");
        }
        return this.client.rest.applications.editGuildCommandPermissions(this.applicationID, this.guildID, this.id, options);
    }
    /**
     * Get this command's permissions (guild commands only).
     */
    async getGuildPermission() {
        if (!this.guildID) {
            throw new TypeError("getGuildPermission cannot be used on global commands.");
        }
        return this.client.rest.applications.getGuildPermission(this.applicationID, this.guildID, this.id);
    }
    /**
     * Get a mention for this command.
     * @param sub The subcommand group and/or subcommand to include (["subcommand"] or ["subcommand-group", "subcommand"]).
     */
    mention(sub) {
        let text = `${this.name}`;
        if (sub?.length) {
            text += ` ${sub.slice(0, 2).join(" ")}`;
        }
        return `</${text}:${this.id}>`;
    }
    toJSON() {
        return {
            ...super.toJSON(),
            applicationID: this.applicationID,
            contexts: this.contexts,
            defaultMemberPermissions: this.defaultMemberPermissions?.toJSON(),
            description: this.description,
            descriptionLocalizations: this.descriptionLocalizations,
            dmPermission: this.dmPermission,
            guildID: this.guildID ?? undefined,
            handler: this.handler,
            integrationTypes: this.integrationTypes,
            name: this.name,
            nameLocalizations: this.nameLocalizations,
            nsfw: this.nsfw,
            options: this.options,
            type: this.type,
            version: this.version
        };
    }
}
exports.default = ApplicationCommand;
