// oceanic.js/dist/lib/structures/Soundboard.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const tslib_1 = require("tslib");
/** @module Soundboard */
const Base_1 = tslib_1.__importDefault(require("./Base"));
const User_1 = tslib_1.__importDefault(require("./User"));
/** Represents a soundboard. */
class Soundboard extends Base_1.default {
    /** If the soundboard sound can be used. */
    available;
    /** The emoji id of the soundboard sound. */
    emojiID;
    /** The emoji name of the soundboard sound. */
    emojiName;
    /** The guild this soundboard sound is in. */
    guildID;
    /** The name of the soundboard sound. */
    name;
    /** The id of the soundboard sound. */
    soundID;
    /** The user who created the soundboard sound. */
    user;
    /** The volume of the soundboard sound. */
    volume;
    constructor(data, client) {
        super(data.sound_id, client);
        this.available = data.available;
        this.emojiID = data.emoji_id;
        this.emojiName = data.emoji_name;
        this.guildID = data.guild_id;
        this.name = data.name;
        this.soundID = data.sound_id;
        this.user = data.user ? new User_1.default(data.user, client) : undefined;
        this.volume = data.volume;
    }
    /**
     * Delete this soundboard sound.
     * @param reason The reason for deleting the soundboard sound.
     */
    async delete(reason) {
        return this.client.rest.guilds.deleteSoundboardSound(this.guildID, this.id, reason);
    }
    /**
     * Edit this soundboard sound.
     * @param options The options for editing the soundboard sound.
     */
    async edit(options) {
        return this.client.rest.guilds.editSoundboardSound(this.guildID, this.id, options);
    }
    /**
     * Send this soundboard sound to a voice channel.
     * @param channelID The ID of the voice channel to send the soundboard sound to.
     * @param sourceGuildID The ID of the guild the soundboard sound is from.
     */
    async sendSoundboardSound(channelID, sourceGuildID) {
        return this.client.rest.channels.sendSoundboardSound(channelID, { soundID: this.soundID, sourceGuildID });
    }
    toJSON() {
        return {
            ...super.toJSON(),
            available: this.available,
            emojiID: this.emojiID,
            emojiName: this.emojiName,
            guildID: this.guildID,
            name: this.name,
            soundID: this.soundID,
            user: this.user?.toJSON(),
            volume: this.volume
        };
    }
}
exports.default = Soundboard;
