// oceanic.js/dist/lib/structures/Subscription.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const tslib_1 = require("tslib");
const Base_1 = tslib_1.__importDefault(require("./Base"));
class Subscription extends Base_1.default {
    /** When the subscription was canceled. */
    canceledAt;
    /** ISO3166-1 alpha-2 country code of the payment source used to purchase the subscription. Missing unless queried with a private OAuth scope. */
    country;
    /** End of the current subscription period. */
    currentPeriodEnd;
    /** Start of the current subscription period. */
    currentPeriodStart;
    /** List of entitlements granted for this subscription. */
    entitlementIDs;
    /** List of SKUs that this user will be subscribed to at renewal. */
    renewalSKUIDs;
    /** List of SKUs subscribed to. */
    skuIDs;
    /** Current status of the subscription. */
    status;
    /**	ID of the user who is subscribed. */
    userID;
    constructor(data, client) {
        super(data.id, client);
        this.canceledAt = data.canceled_at ? new Date(data.canceled_at) : null;
        this.country = data.country;
        this.currentPeriodEnd = new Date(data.current_period_end);
        this.currentPeriodStart = new Date(data.current_period_start);
        this.entitlementIDs = data.entitlement_ids;
        this.renewalSKUIDs = data.renewal_sku_ids;
        this.skuIDs = data.sku_ids;
        this.status = data.status;
        this.userID = data.user_id;
    }
}
exports.default = Subscription;
