// oceanic.js/dist/lib/routes/Miscellaneous.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const tslib_1 = require("tslib");
/** @module REST/Miscellaneous */
const Routes = tslib_1.__importStar(require("../util/Routes"));
const Soundboard_1 = tslib_1.__importDefault(require("../structures/Soundboard"));
/** Methods that don't fit anywhere else. Located at {@link Client#rest | Client#rest}{@link RESTManager#misc | .misc}. */
class Miscellaneous {
    _manager;
    constructor(manager) {
        this._manager = manager;
    }
    /**
     * Get the default soundboard sounds.
     * @caching This method **does not** cache its result.
     */
    async getDefaultSoundboardSounds() {
        return this._manager.authRequest({
            method: "GET",
            path: Routes.SOUNDBOARD_DEFAULT_SOUNDS
        }).then(data => data.map(d => new Soundboard_1.default(d, this._manager.client)));
    }
    /**
     * Get a sticker.
     * @param stickerID The ID of the sticker to get.
     * @caching This method **may** cache its result. The result will not be cached if the guild is not cached, or if the sticker is not a guild sticker.
     * @caches {@link Guild#stickers | Guild#stickers}
     */
    async getSticker(stickerID) {
        return this._manager.authRequest({
            method: "GET",
            path: Routes.STICKER(stickerID)
        }).then(data => data.guild_id === undefined ? this._manager.client.util.convertSticker(data) : this._manager.client.guilds.get(data.guild_id)?.stickers.update(data) ?? this._manager.client.util.convertSticker(data));
    }
    /**
     * Get the default sticker packs.
     * @caching This method **does not** cache its result.
     */
    async getStickerPacks() {
        return this._manager.authRequest({
            method: "GET",
            path: Routes.STICKER_PACKS
        }).then(data => data.sticker_packs.map(pack => ({
            bannerAssetID: pack.banner_asset_id,
            coverStickerID: pack.cover_sticker_id,
            description: pack.description,
            id: pack.id,
            name: pack.name,
            skuID: pack.sku_id,
            stickers: pack.stickers.map(sticker => this._manager.client.util.convertSticker(sticker))
        })));
    }
    /**
     * Get the list of usable voice regions.
     * @caching This method **does not** cache its result.
     */
    async getVoiceRegions() {
        return this._manager.authRequest({
            method: "GET",
            path: Routes.VOICE_REGIONS
        });
    }
    /**
     * Refresh expired attachment URLs.
     * @param urls The CDN urls to refresh.
     */
    async refreshAttachmentURLs(urls) {
        return this._manager.authRequest({
            method: "POST",
            path: Routes.REFRESH_ATTACHMENT_URLS,
            json: {
                attachment_urls: urls
            }
        }).then(data => ({
            refreshedURLs: data.refreshed_urls
        }));
    }
}
exports.default = Miscellaneous;
